### ISSUE TITLE
Feedback on latest Editor's Draft

### ISSUE THREAD

**Issue by benfrancis** [w3c/wot-discovery#299]
As with the Thing Description specification in https://github.com/w3c/wot-thing-description/issues/1324 I've carried out a full review of the latest Editor's Draft of the WoT Discovery specification prior to normative feature freeze, with a view to implementation in WebThings.

I hope this feedback is helpful. Sorry to include all of this in one big issue but there are a lot of minor typo corrections and editorial changes which don't necessarily justify their own issues. I'm happy to file separate issues for the bigger points if needed.

Most of my feedback is just editorial suggestions to make the specification more readable. Apart from some niggles regarding HTTP headers, content types and encryption requirements, my one big concern is the lack of a defined authentication mechanism for HTTP-based exploration mechanisms. In my view, until issue #135 is resolved, the discovery process is largely useless for ad-hoc discovery because the two step discovery process assumes that a Consumer already has authentication credentials for a device before it discovers it. I think this needs resolving prior to publication. There's a proposed solution in #135 but I'm waiting on feedback.

Reminder: There are quite a few remaining Editor's Notes and Issues which need resolving before publication.

# 3. Terminology 

- [ ]  The [wot-architecture] reference needs updating to [wot-architecture11] where the new terms are defined.
- [ ] The definition of the term "Discoverer" in WoT Architecture 1.1 appears to be inconsistent with the way the term is used in WoT Discovery 1.1. See https://github.com/w3c/wot-architecture/issues/738
- [ ] I think the specification could be made clearer by consistently using the following terms in the prose of the text:
  * Thing
  * Directory (aka Thing Directory or Thing Description Directory)
  * Thing Description
  * Directory Description (aka Thing Directory Description)
(Otherwise the distinction between a Thing Description (TD), Thing Directory (TD), Thing Directory Description (TDD) and Thing Description Directory (TDD) gets very confusing.)
- [ ] Note that there is currently not a definition of "Directory Description", though the term is used in several places.

# 4. Architecture

- [ ] Figure 1 seems to indicate that a directory is the only exploration mechanism, it isn't completely clear that self-description can be an exploration mechanism too. Also the section numbers are out of date.

# 5. Discoverer Process

- [ ] As noted in https://github.com/w3c/wot-architecture/issues/738 I think the term "Discoverer" is a bit confusing. Also, this section doesn't really define a process so much as a set of requirements or "responsibilities" of a "Discoverer" entity. An alternative title for this section might therefore be "Discovery Agent", "Discovery Consumer" or "Discovery Client".
- [ ] Typo: "A Discoverer MUST support fetching a TD from least one URL provided as part of the Introduction process." -> "A Discoverer MUST support fetching a TD from at least one URL provided as part of the Introduction process."

## 6.1 Direct

> A request on all such URLs MUST result in a TD as prescribed in [7.1 Self-description](https://w3c.github.io/wot-discovery/#exploration-self).

- [ ] When I read the specification from top to bottom, I found it a bit confusing that this assertion referenced section 7.1 but not 7.2. It's only after reading section 7 as a whole that it becomes clear that a Directory can itself be a Thing which self-describes itself. Might it be worth referencing section 7 as a whole here rather than 7.1 to avoid the impression that the Direct introduction mechanism can only be used for self-describing things and not directories?

## 6.2 Well-Known URIs

- [ ] Suggest re-wording the following assertion slightly: "If the Thing or Thing Description Directory use the Well-Known Uniform Resource Identifier to advertise its presense, it MUST register its own Thing or Directory Description into the following path" -> "If the Thing or Thing Description Directory uses a Well-Known Uniform Resource Identifier [RFC8615] to advertise its presense, it MUST register its own Thing Description or Directory Description into the following path"

> When a request is made at the above Well-Known URI, the server MUST return a Thing Description as prescribed in 7.1 Self-description.

- [ ] Similar to section 6.1, it might be better to reference section 7 as a whole here rather than section 7.1.

## 6.3 DNS-Based Service Discovery

- [ ] "A Thing or Thing Description Directory MAY use the DNS-Based Service Discovery (DNS-SD)[RFC6763]." -> "A Thing or Thing Description Directory MAY use DNS-Based Service Discovery (DNS-SD)[RFC6763]."

> This can be also be used to discover them on the same link by combining Multicast DNS (mDNS)[RFC6762].

- [ ] What does "link" mean? Should this say "local network"?
- [ ] "In DNS-SD, format of the Service Instance Name is Instance.Service.Domain." -> "In DNS-SD, the format of the Service Instance Name is Instance.Service.Domain."
- [ ] "The Service Name to indicate the Thing or Thing Description Directory MUST be _wot. And the Service Name to indicate the Thing Description Directory MUST be _directory._sub._wot." -> "The Service Name to indicate the Thing or Thing Description Directory MUST be _wot and the Service Name to indicate the Thing Description Directory MUST be _directory._sub._wot."

## 6.5 DID Documents

- [ ] "The DID Document obtained by resolving the DID of a Thing or Thing Description Directory MUST contains a Service Endpoint which point to Thing Description of the Thing or Directory Description of the Thing Description Directory." -> "The DID Document obtained by resolving the DID of a Thing or Thing Description Directory MUST contain a Service Endpoint which points to a Thing Description of the Thing or Directory Description of the Thing Description Directory."

# 7 Exploration Mechanisms

I am still a bit sceptical as to whether Thing Links are actually necessary, because the example scenarios given in this section don't seem very compelling to me. I struggle to think of practical use cases where a Thing Link would be used where the full Thing Description couldn't just be used instead. I guess we need to wait and see what level of adoption this gets by implementors.

## 7.1 Self-description

> The server SHOULD serve the requests after performing necessary authentication and authorization.
...
> The HTTP server MUST serve the TD with a GET method. A successful response MUST have 200 (OK) status, contain application/td+json Content-Type header, and the TD in body.

- [ ] The specification really needs to define how authentication and authorisation works for HTTP. When a Consumer is provided with a URL from an introduction mechanism, it isn't provided with authentication credentials. When it performs an initial `GET` request on the URL it is therefore likely to get a `401 Unauthorized` response rather than a `200 OK` response. The specification says that a Consumer may receive a `401 Unauthorized` or `403 Forbidden` response, but it doesn't define what it should do when this happens, despite it likely being the most common case. See #135.

## 7.2 Directory

- [ ] I think this section needs an explanation that a Directory is described by a Directory Description, with an example of a Directory Description. There used to be an example Thing Description of a Directory in section 7, but it was turned into a Thing Model at moved to the end in section 7.2.2.4. I think an example Thing Description would be useful.

- [ ] I find the use of the terms "client", "server", "consumer" and "producer" confusing throughout this section. It's not always clear whether client/consumer refers to a consumer of a Thing Description or a consumer of the Thing Directory API and it's not clear whether server/producer refers to a self-describing Thing or the Directory Service API server. It would be useful to define these terms up-front.

I think what is intended is:
* Client - A client of the Directory Service API
* Server - The Directory Service API server
* Consumer - A consumer of a Thing Description
* Producer - A producer of a Thing Description, e.g. a self-describing Thing

Possible alternative terms might be "Directory Client" and "Directory Server", or "Directory Consumer" and "Directory Producer", to disambiguate from "Thing Description Consumer" and "Thing Description Producer".

#### 7.2.1.1 Registration Information

As an example of the above, does the `retrieved` member indicate when a Thing Description was retrieved from a self-describing Thing, or when a Consumer of the Directory Service API retrieved the Thing Description from the directory? The term "server" is a bit ambiguous here.

#### 7.2.1.2 Registration Expiry

- [ ] "Section 7.2.1.1 Registration Information introduces few attributes to specify and discover the expiry time of registered TDs." -> "Section 7.2.1.1 Registration Information introduces some attributes to specify and discover the expiry time of registered TDs."
- [ ]  This section defines how a registration expires, but it doesn't describe how to prevent it expiring. Is a registrant expected to re-register the Thing Description periodically using the create or update methods of the Things API, or is the directory expected to re-retrieve an up to date Thing Description from somewhere for example?

### 7.2.2 Directory Service API

> The HTTP API MUST be exposed over HTTPS (HTTP Over TLS). 

- [ ] It's not possible to conform with this assertion if a directory is exposed over a local network, without using self-signed certificates, which would then cause problems for clients (e.g. clients written in client-side JavaScript because the certificate will generate a security exception in browsers). This will be a problem for WebThings Gateway which always exposes its directory over a local .local domain, but only optionally over a global .io domain. As with may other cases in WoT where we've discussed making TLS mandatory, I don't think this can be enforced, and it really has to be a SHOULD rather than a MUST.
- [ ] Similar to section 7.1, there are many instances throughout this section where 401 Unauthorized and 403 Forbidden responses are used, but there's no explanation about what a client should do if it receives this response. Where does it get authentication credentials from? See #135.

##### 7.2.2.1.2 Retrieval

- [ ] I noticed that for GET requests, the response is expected to have a `Content-Type` header of `application/td+json` but no `Accept` header is sent to request a resource in that format. This will cause a problem for WebThings Gateway because the same URLs are used for both a JSON representation of a Thing and an HTML representation of a Thing. By default the HTML representation is returned. Would it be possible to require an `Accept` header be set here with a value of `application/td+json`?

##### 7.2.2.1.5 Listing

- [ ] Same as 7.2.2.1.2, would it be possible to require an `Accept` header here to ask for a JSON representation? WebThings Gateway will currently provide an HTML representation if this is not provided.
- [ ] I note that the response for this operation uses the `application/ld+json` content-type rather than `application/td+json` because the resource it returns is not a valid Thing Description, it's a list of Thing Descriptions. Could `application/json` be acceptable here too, as is used in the Search API?
- [ ] "Alternative to a pure array of TDs as the body of the response, the server MAY send a more verbose payload allowing server-side information, like pagination information, to be put in addition to the actual data." -> "As an alternative to an array of TDs as the body of the response, the server MAY send a more verbose payload allowing server-side information, like pagination information, to be included in addition to the actual data."
- [ ]  "it looks like as follows for the listing endpoint" -> "it looks as follows for the listing endpoint"

#### 7.2.2.2 Events API

> The server SHOULD provide an event ID as the id field in each event and respond to re-connecting clients by delivering all missed events. 

Just FYI, in the HTTP SSE Profile we're discussing recommending or mandating that IDs be timestamps. See https://github.com/w3c/wot-profile/issues/100.

#### 7.2.2.4 API Specification (Thing Model)

- [ ] I've noticed that on the forms for events there is a `contentType` set on requests with a value of `text/event-stream`, but I think this is incorrect because in SSE it's actually the response which has this content type, not the request. You therefore need to set an `Accept` header with a value of `text/event-stream` instead, and the `Content-type` header of the response should then have a value of `text/event-stream`.  For SSE you also need to set the `Connection` header to `keep-alive`, but I don't know whether this is implied by the subprotocol. See below for example metadata:

```
"htv:headers": [
  {
    "htv:fieldName": "Accept",
    "htv:fieldValue": "text/event-stream"
  },
  {
    "htv:fieldName": "Connection",
    "htv:fieldValue": "keep-alive"
  }
],
"response": {
  "htv:statusCodeNumber": 200,
  "contentType": "text/event-stream"
}
```

- [ ] We might also want to consider adding a top level form for a `subscribeallevents` operation to describe the /events endpoint, since that has now been added to the Thing Description specification. I've included an example below:

```
{
  "op": "subscribeallevents",
  "href": "/events",
  "subprotocol": "sse",
  "htv:methodName": "GET",
  "htv:headers": [
    {
      "htv:fieldName": "Accept",
      "htv:fieldValue": "text/event-stream"
    },
    {
      "htv:fieldName": "Connection",
      "htv:fieldValue": "keep-alive"
    }
  ],
  "response": {
    "htv:statusCodeNumber": 200,
    "contentType": "text/event-stream"
  }
},
```

**Comment by mmccool** [comment]
Ben, thanks for this thorough review!  We discussed this in the Discovery call 2022.05.02.  Following are our actions items to resolve:
3. @mmccool is working on these, as part of general cleanup in coordinating terminology, fixing definition of Discoverer term, etc.  For now I will focus on making the use of the term "Discoverer" consistent rather than trying to edit in a new term.
4. @mmccool will look at the figure again and attempt to make this clearer
5. @mmccool to work on cleanup, may change section title to be clearer (but not "Discoverer" term itself, the suggested alternatives all have their own problems...)
6. @k-toumura will address
7. @farshidtz (with assistance from @AndreaCimminoArriaga) will see what he can do to address most of these; as far as security bootstrapping, I think the answer should be that the service should return 401 Unauthorized, but with information in the HTTP protocol on HOW to obtain authorization... in other words, negotiation, just like a regular web service.  Then the client should attempt to reconnect with the appropriate credentials (if it has them, which it will if it is authorized).
8.

**Comment by mmccool** [comment]
> The definition of the term "Discoverer" in WoT Architecture 1.1 appears to be inconsistent with the way the term is used in WoT Discovery 1.1. See https://github.com/w3c/wot-architecture/issues/738
See https://github.com/w3c/wot-architecture/issues/738, which defines both "Discoverer" and "Registrant".  Note: I intend to add some text to mention "Registrant" soon.

PR #xxx addresses the following sub-issues as stated:
> The [wot-architecture] reference needs updating to [wot-architecture11] where the new terms are defined.
Actually, this reference should be updated generally to 1.1 and the reference to "wot-thing-description" should also be updated to 1.1.  Fixed.  

> Typo: "A Discoverer MUST support fetching a TD from least one URL provided as part of the Introduction process." -> "A Discoverer MUST support fetching a TD from at least one URL provided as part of the Introduction process."
Suggestion adopted.

> When I read the specification from top to bottom, I found it a bit confusing that this assertion referenced section 7.1 but not 7.2. It's only after reading section 7 as a whole that it becomes clear that a Directory can itself be a Thing which self-describes itself. Might it be worth referencing section 7 as a whole here rather than 7.1 to avoid the impression that the Direct introduction mechanism can only be used for self-describing things and not directories?
Changed to reference section 7 as a whole.

> Suggest re-wording the following assertion slightly: "If the Thing or Thing Description Directory use the Well-Known Uniform Resource Identifier to advertise its presense, it MUST register its own Thing or Directory Description into the following path" -> "If the Thing or Thing Description Directory uses a Well-Known Uniform Resource Identifier [RFC8615] to advertise its presense, it MUST register its own Thing Description or Directory Description into the following path"
Added the RFC reference to the assertion.  Also spelled "presense" as "presence".

> Similar to section 6.1, it might be better to reference section 7 as a whole here rather than section 7.1.
Not as clear-cut as the previous case, but sure, changed.

> "A Thing or Thing Description Directory MAY use the DNS-Based Service Discovery (DNS-SD)[RFC6763]." -> "A Thing or Thing Description Directory MAY use DNS-Based Service Discovery (DNS-SD)[RFC6763]."
This is one of those "grammar makes sense if you expand the acronym" problem.  I have deleted the "the" although another solution might be to use "the DNS-Based Service Discovery (DNS-SD)[RFC6763] mechanism" which is grammatically correct whether or not you expand the acronym.  However, leaving out the "the" is less verbose so let's do that.

> What does "link" mean? Should this say "local network"?
Reworded to say local network.  Actually this whole sentence was weirdly worded so I reworded it, hopefully less weirdly.

>"In DNS-SD, format of the Service Instance Name is Instance.Service.Domain." -> "In DNS-SD, the format of the Service Instance Name is Instance.Service.Domain."
Fixed.

>  "The Service Name to indicate the Thing or Thing Description Directory MUST be _wot. And the Service Name to indicate the Thing Description Directory MUST be _directory._sub._wot." -> "The Service Name to indicate the Thing or Thing Description Directory MUST be _wot and the Service Name to indicate the Thing Description Directory MUST be _directory._sub._wot."
Actually, the reason for the weird "And" here is that they are two separate assertions.  Rather than merging them into one sentence as suggested, I reworded the second sentence so it is grammatically correct: "And the" -> "The".

(more to follow, going to merge the PR now, then keep working on it)

**Comment by mmccool** [comment]
Ooops, looks like I should have left section 6 alone as @k-toumura was supposed to work on it.  Sorry... well, we can sort it out.

**Comment by mmccool** [comment]
Also attempting to address the TLS issue with some general rules about public/private networks, and PII risks.  Some of this is also to address feedback from PING.  See the following: https://github.com/w3c/wot-architecture/pull/747
This PR is in architecture but once we can establish the general principles, we can perhaps add some more specific assertions to the Discovery spec.  In this case the assertion about TLS in 7.2.2 would have to become a little more refined, either referring to the Architecture security section OR explictly expanding into two assertions (MUST for public, SHOULD for private).   We probably also should add an assertion disallowing nosec in cases where PII is involved or IDs are immutable (see also above PR).

**Comment by farshidtz** [comment]
> #### 7.2.1.1 Registration Information
> As an example of the above, does the `retrieved` member indicate when a Thing Description was retrieved from a self-describing Thing, or when a Consumer of the Directory Service API retrieved the Thing Description from the directory? The term "server" is a bit ambiguous here.
> 

This is currently explained above the table:
_In this table, client refers to the producer or consumer of a [TD](https://w3c.github.io/wot-discovery/#dfn-td) and server refers to the [Thing Description Directory](https://w3c.github.io/wot-discovery/#dfn-wot-thing-description-directory)._

> ##### 7.2.2.1.2 Retrieval
> * [ ]  I noticed that for GET requests, the response is expected to have a `Content-Type` header of `application/td+json` but no `Accept` header is sent to request a resource in that format. This will cause a problem for WebThings Gateway because the same URLs are used for both a JSON representation of a Thing and an HTML representation of a Thing. By default the HTML representation is returned. Would it be possible to require an `Accept` header be set here with a value of `application/td+json`?
> 

See https://w3c.github.io/wot-discovery/#exploration-directory-api-things

- _The default serialization format for all request and success response bodies MUST be JSON, with JSON-LD 1.1 [[JSON-LD11](https://w3c.github.io/wot-discovery/#bib-json-ld11)] syntax to support extensions and semantic processing._
- _Directories MAY accept additional representations based on request's indicated Content-Type or Content-Encoding, and provide additional representations through server-driven content negotiation._

The directory must use JSON format as default. It may accept other formats as input and respond other formats as output. I think we need to have a default rather than requiring `Accept` in all submit requests.

> ##### 7.2.2.1.5 Listing
> * [ ]  Same as 7.2.2.1.2, would it be possible to require an `Accept` header here to ask for a JSON representation? WebThings Gateway will currently provide an HTML representation if this is not provided.

Same as above. The directory can react to the `Accept` header and if possible give another representation, but the default must be JSON. Perhaps create a separate issue if you'd like to propose having another default or no default at all.

> * [ ]  I note that the response for this operation uses the `application/ld+json` content-type rather than `application/td+json` because the resource it returns is not a valid Thing Description, it's a list of Thing Descriptions. Could `application/json` be acceptable here too, as is used in the Search API?

Assertions are:
- _A successful response MUST have 200 (OK) status, contain application/ld+json Content-Type header, and an array of TDs in the body._ [listing](https://w3c.github.io/wot-discovery/#exploration-directory-api-things-listing)
- _A successful response MUST have 200 (OK) status, contain application/td+json Content-Type header, and the requested TD in body._ [retrieval](https://w3c.github.io/wot-discovery/#exploration-directory-api-things-retrieval)

I think we can relax the content types to recommendations. IMO, the `application/json` Content-Type are acceptable for both retrieval and listing.

**Comment by mmccool** [comment]
My plans for next week:
- Clean up the "Discovery Process" section to resolve terminology issues
- Add a PR to address #135 - based on use of 401 errors to initiate negotiation in specific cases

**Comment by benfrancis** [comment]
@relu91 wrote:
> I think a protocol binding template document would be better suited for such an explanation.

As I understand it protocol binding templates currently only apply to interaction affordances, where protocol bindings are serialised as Forms inside a Thing Description. This issue is about authenticating access to the Thing Description itself. That's why I think this needs to be covered in WoT Discovery, which defines mechanisms for discovering and retrieving Thing Descriptions.
